File: 085_Maximal_Rectangle/085_1.c
int maximalRectangle(char** matrix, int matrixSize, int* matrixColSize){
    if (matrixSize == 0 || *matrixColSize == 0) return 0;
    int res = 0;
    int m = matrixSize, n = *matrixColSize;
    int height[n + 2];
    memset(height, 0, sizeof(height));
    for (int i = 0; i < m; i++) {
        int stack[n + 2];
        memset(stack, 0, sizeof(stack));
        int top = 0;
        for (int j = 1; j < n + 2; j++) {
            if (j-1 < n && matrix[i][j-1] == '1') {
                height[j]++;
            } else {
                height[j] = 0;
            }
            while (height[stack[top]] > height[j]) {
                int curHeight = height[stack[top]];
                top--;
                int curWidth = j - stack[top] - 1;
                res = fmax(res, curHeight * curWidth);
            }
            stack[++top] = j;
        }
    }
    return res;
}

File: 085_Maximal_Rectangle/085_1.py
from typing import List

class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        
        res, m, n = 0, len(matrix), len(matrix[0])
        height = [0 for _ in range(n+2)]
        for i in range(m):
            stack = [0]
            for j in range(1, n+2):
                if j-1 < n and matrix[i][j-1] == '1':
                    height[j] += 1
                else:
                    height[j] = 0
                while height[stack[-1]] > height[j]:
                    curHeight = height[stack.pop()]
                    curWidth = j - stack[-1] - 1
                    res = max(res, curHeight * curWidth)
                stack.append(j)
        return res
